src/main/java/org/sagebionetworks/bridge/hibernate/HibernateSurveyDao.java

```diff
@@ -0,0 +1,365 @@
+    }
+    
+    @Autowired
+    public final void setUploadSchemaService(UploadSchemaService uploadSchemaService) {
+        this.uploadSchemaService = uploadSchemaService;
+    }
+    
+    @Override
+    public Survey createSurvey(Survey survey) {
+        System.out.println("createSurvey");
+        checkNotNull(survey.getStudyIdentifier(), "Survey study identifier is null");
+        if (survey.getGuid() == null) {
+            survey.setGuid(generateGuid());
+        }
+        long time = DateUtils.getCurrentMillisFromEpoch();
+        survey.setCreatedOn(time);
+        survey.setModifiedOn(time);
+        survey.setSchemaRevision(null);
+        survey.setPublished(false);
+        survey.setDeleted(false);
+        survey.setVersion(null);
+        hibernateHelper.create(survey, null);
+        System.out.println("12");
+        return survey;
+    }
+
+    @Override
+    public Survey updateSurvey(StudyIdentifier studyIdentifier, Survey survey) {
+        Survey existingSurvey = getSurvey(studyIdentifier, survey, true);
+        
+        if (existingSurvey != null) {
+            hibernateHelper.update(survey, null);
+            return survey;
+        } else {
+            return createSurvey(survey);
+        }
+    }
+
+    @Override
+    public Survey versionSurvey(StudyIdentifier studyIdentifier, GuidCreatedOnVersionHolder keys) {
+        Survey existingSurvey = getSurvey(studyIdentifier, keys, false);
+        System.out.println(existingSurvey);
+        HibernateSurvey copy = new HibernateSurvey(existingSurvey);
+        copy.setSchemaRevision(null);
+        for (SurveyElement element : copy.getElements()) {
+            element.setGuid(generateGuid());
+        }
+        System.out.println("12");
+        return createSurvey(copy);
+    }
+
+    @Override
+    public Survey publishSurvey(StudyIdentifier studyIdentifier, Survey survey, boolean newSchemaRev) {        
+        if (!survey.isPublished()) {
+            // update survey
+            survey.setPublished(true);
+            survey.setModifiedOn(DateUtils.getCurrentMillisFromEpoch());
+            // make schema from survey
+            if (!survey.getUnmodifiableQuestionList().isEmpty()) {
+                UploadSchema schema = uploadSchemaService.createUploadSchemaFromSurvey(studyIdentifier, survey, newSchemaRev);
+                survey.setSchemaRevision(schema.getRevision());
+            }
+            updateSurvey(studyIdentifier, survey);
+        }
+        return survey;
+    }
+
+    @Override
+    public void deleteSurvey(Survey survey) {
+        survey.setDeleted(true);
+        hibernateHelper.update(survey, null);
+    }
+    
+    @Override
+    public void deleteSurveyPermanently(StudyIdentifier studyIdentifier, GuidCreatedOnVersionHolder keys) {
+        Survey existing = getSurvey(studyIdentifier, keys, false);
+        
+        if (existing != null) {
+            deleteAllElements(existing.getGuid(), existing.getCreatedOn());
```

Comment:
If we use "on cascade delete" for the foreign keys, is it really necessary to explicitly delete the survey elements before deleting the survey?